source/common/upstream/cluster_manager_impl.cc

```diff
@@ -313,12 +313,25 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   // Now setup for cross-thread updates.
   cluster.prioritySet().addMemberUpdateCb(
       [&cluster, this](const HostVector&, const HostVector& hosts_removed) -> void {
-        // TODO(snowp): Should this be subject to merge windows?
+        const bool close_connections_on_host_set_change =
+            cluster.info()->lbConfig().close_connections_on_host_set_change();
 
-        // Whenever hosts are removed from the cluster, we make each TLS cluster drain it's
-        // connection pools for the removed hosts.
-        if (!hosts_removed.empty()) {
-          postThreadLocalHostRemoval(cluster, hosts_removed);
+        if (close_connections_on_host_set_change) {
+          for (const auto& host_set : cluster.prioritySet().hostSetsPerPriority()) {
+            // this will drain all tcp and http connection pools
+            postThreadLocalHostRemoval(cluster, host_set->hosts());
```

Comment:
For the sake of code clarity, do you mind updating `postThreadLocalHostRemoval`to make it clear that it doesn't remove the hosts from the TLS cluster, just drains its connections?

Reply:
@snowp i think that in terms of `ThreadLocalClusterManagerImpl` `postThreadLocalHostRemoval` actually removes the host. TLS cluster manager stores mappings for tcp and http connections in form of `host_ptr->connection_pool_container`. So whenever connection pool is drained, corresponding entry is erased from map and tls manager has no longer any knowledge about the host. Name sounds reasonable to me.

Reply:
it removes it from the connection pools, but not from the TLS `PrioritySet`. I had to go check the implementation of `postThreadLocalHostRemoval` to make sure that we weren't doing a "remove -> add back in" of existing endpoints (which would have resulting in intermittent UH during host selection). Maybe something like `postThreadLocalDrainConnections` would make it clearer that it's not touching the load balancing structures?

Reply:
ok, I see the point. Missed local_priority_set on TLS clusters. Renaming to `postThreadLocalDrainConnections`